crypto/src/main/java/org/web3j/crypto/StructuredDataEncoder.java

```diff
@@ -0,0 +1,450 @@
+                    // Skip types which are already expanded
+                    continue;
+                } else {
+                    // Encountered a user defined type
+                    remainingTypes.add(entry.getType());
+                }
+            }
+        }
+
+        return deps;
+    }
+
+    public String encodeStruct(String structName) {
+        HashMap<String, List<StructuredData.Entry>> types = jsonMessageObject.getTypes();
+
+        String structRepresentation = structName + "(";
+        for (StructuredData.Entry entry: types.get(structName)) {
+            structRepresentation += String.format("%s %s,", entry.getType(), entry.getName());
+        }
+        structRepresentation = structRepresentation.substring(0, structRepresentation.length() - 1);
+        structRepresentation += ")";
+
+        return structRepresentation;
+    }
+
+    public String encodeType(String primaryType) {
+        HashSet<String> deps = getDependencies(primaryType);
+        deps.remove(primaryType);
+
+        // Sort the other dependencies based on Alphabetical Order and finally add the primaryType
+        List<String> depsAsList = new ArrayList<>(deps);
+        Collections.sort(depsAsList);
+        depsAsList.add(0, primaryType);
+
+        String result = "";
+        for (String structName: depsAsList) {
+            result += encodeStruct(structName);
+        }
+
+        return result;
+    }
+
+    public byte[] typeHash(String primaryType) {
+        return Numeric.hexStringToByteArray(sha3String(encodeType(primaryType)));
+    }
+
+    public List<Integer> getArrayDimensionsFromDeclaration(String declaration) {
+        Pattern arrayTypePattern = Pattern.compile(arrayTypeRegex);
+
+        // This regex tries to extract the dimensions from the
+        // square brackets of an array declaration using the ``Regex Groups``.
+        // Eg- It extracts ``5, 6, 7`` from ``[5][6][7]``
+        String arrayDimensionRegex = "\\[([1-9]\\d*)?\\]";
+        Pattern arrayDimensionPattern = Pattern.compile(arrayDimensionRegex);
+
+        // Get the dimensions which were declared in Schema.
+        // If any dimension is empty, then it's value is set to -1.
+        Matcher arrayTypeMatcher = arrayTypePattern.matcher(declaration);
+        arrayTypeMatcher.find();
+        String dimensionsString = arrayTypeMatcher.group(1);
+        Matcher dimensionTypeMatcher = arrayDimensionPattern.matcher(dimensionsString);
+        List<Integer> dimensions = new ArrayList<>();
+        while (dimensionTypeMatcher.find()) {
+            String currentDimension = dimensionTypeMatcher.group(1);
+            if (currentDimension == null) {
+                dimensions.add(Integer.parseInt("-1"));
+            } else {
+                dimensions.add(Integer.parseInt(currentDimension));
+            }
+        }
+
+        return dimensions;
+    }
+
+    public ArrayList<Pair> getDepthsAndDimensions(Object data, int depth) {
```

Comment:
Can you use a map here?

Reply:
Or even create your own `Pair` class if there isn't already one on the classpath outside java.fx?